fluid/voice.cpp

```diff
@@ -326,78 +371,110 @@ void Voice::write(unsigned n, float* out, float* reverb, float* chorus)
+                  /* the envelope is in the attack section: ramp linearly to max value.
+                   * A positive modlfo_to_vol should increase volume (negative attenuation).
+                   */
+                  target_amp = fluid_atten2amp (attenuation)
+                     * fluid_cb2amp (modlfo_val * -modlfo_to_vol)
+                     * volenv_val;
+                  }
+            else {
+                  //float amplitude_that_reaches_noise_floor;
+                  //float amp_max;
 
-            /* voice->attenuation_min is a lower boundary for the attenuation
-             * now and in the future (possibly 0 in the worst case).  Now the
-             * amplitude of sample and volenv cannot exceed amp_max (since
-             * volenv_val can only drop):
-             */
+                  target_amp = fluid_atten2amp (attenuation)
+                     * fluid_cb2amp (960.0f * (1.0f - volenv_val)
+                     + modlfo_val * -modlfo_to_vol);
+
+                  /* We turn off a voice, if the volume has dropped low enough. */
+
+                  /* A voice can be turned off, when an estimate for the volume
+                   * (upper bound) falls below that volume, that will drop the
+                   * sample below the noise floor.
+                   */
+
+                  /* If the loop amplitude is known, we can use it if the voice loop is within
+                   * the sample loop
+                   */
+
+                  /* Is the playing pointer already in the loop? */
+                  /*if (has_looped)
+                        amplitude_that_reaches_noise_floor = amplitude_that_reaches_noise_floor_loop;
+                  else
+                        amplitude_that_reaches_noise_floor = amplitude_that_reaches_noise_floor_nonloop;*/
+
+                  /* voice->attenuation_min is a lower boundary for the attenuation
+                   * now and in the future (possibly 0 in the worst case).  Now the
+                   * amplitude of sample and volenv cannot exceed amp_max (since
+                   * volenv_val can only drop):
+                   */
 
-            amp_max = fluid_atten2amp (min_attenuation_cB) * volenv_val;
+                  //amp_max = fluid_atten2amp (min_attenuation_cB) * volenv_val;
 
-            /* And if amp_max is already smaller than the known amplitude,
-             * which will attenuate the sample below the noise floor, then we
-             * can safely turn off the voice. Duh. */
-            if (amp_max < amplitude_that_reaches_noise_floor) {
-                  off();
-                  ticks += n;
-                  return;
+                  /* And if amp_max is already smaller than the known amplitude,
+                   * which will attenuate the sample below the noise floor, then we
+                   * can safely turn off the voice. Duh. */
+                  /*if (amp_max < amplitude_that_reaches_noise_floor) {
```

Comment:
Yes please. It's not that it really necessary for now. But it would be good to have a reminder for later that explain what this code was doing, why it's commented and maybe how it could be done again. This one in particular is quite nice.